File: 295-find-median-from-data-stream/find-median-from-data-stream.java
class MedianFinder {

    PriorityQueue<Integer> maxHeap = new PriorityQueue<>((a, b) -> {
        return Integer.compare(b, a);
    });
    PriorityQueue<Integer> minHeap = new PriorityQueue<>();

    public MedianFinder() {
        // do nothing for now
        maxHeap = new PriorityQueue<>((a, b) -> {
            return Integer.compare(b, a);
        });
        minHeap = new PriorityQueue<>();
    }
    
    public void addNum(int num) {
        if (minHeap.isEmpty() && maxHeap.isEmpty()) {
            maxHeap.add(num);
            return;
        }
        if (num < maxHeap.peek()) {
            maxHeap.add(num);
        } else {
            minHeap.add(num);
        }
        if (maxHeap.size() - minHeap.size() > 1) {
            minHeap.add(maxHeap.poll());
        } else if (minHeap.size() - maxHeap.size() > 1) {
            maxHeap.add(minHeap.poll());
        }
    }
    
    public double findMedian() {
        int len1 = maxHeap.size();
        int len2 = minHeap.size();
        if ((len1 + len2) % 2 == 0) {
            return ((double)maxHeap.peek() + (double)minHeap.peek())/2;
        } else {
            if (maxHeap.size() > minHeap.size()) return maxHeap.peek();
            return minHeap.peek();
        }
    }
}

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder obj = new MedianFinder();
 * obj.addNum(num);
 * double param_2 = obj.findMedian();
 */

File: 295-find-median-from-data-stream/find-median-from-data-stream.py
class MedianFinder:

    def __init__(self):
        self.max_heap = []
        self.min_heap = []
        heapify(self.min_heap)
        heapify(self.max_heap)

    def addNum(self, num: int) -> None:
        left_len = len(self.max_heap)
        right_len = len(self.min_heap)
        if left_len == 0:
            heappush(self.max_heap, -num)
            return
        # try putting the num on the min_heap
        if num < -self.max_heap[0]:
            heappush(self.max_heap, -num)
        else:
            heappush(self.min_heap, num)
        # rebalance the heaps
        if abs(len(self.max_heap) - len(self.min_heap)) >= 2:
            left_len = len(self.max_heap)
            right_len = len(self.min_heap)
            if left_len > right_len:
                heappush(self.min_heap, -heappop(self.max_heap))
            else:
                heappush(self.max_heap, -heappop(self.min_heap))      

    def findMedian(self) -> float:
        left_len = len(self.max_heap)
        right_len = len(self.min_heap)
        if (left_len + right_len)%2 == 0:
            return (-self.max_heap[0] + self.min_heap[0]) / 2
        if left_len > right_len:
            return -self.max_heap[0]
        return self.min_heap[0]


# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()